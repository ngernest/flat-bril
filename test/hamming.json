{
  "functions": [
    {
      "args": [
        {
          "name": "val",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "dest": "two",
          "op": "const",
          "type": "int",
          "value": 2
        },
        {
          "args": [
            "val",
            "two"
          ],
          "dest": "div",
          "op": "div",
          "type": "int"
        },
        {
          "args": [
            "div",
            "two"
          ],
          "dest": "mul",
          "op": "mul",
          "type": "int"
        },
        {
          "args": [
            "val",
            "mul"
          ],
          "dest": "lsb",
          "op": "sub",
          "type": "int"
        },
        {
          "args": [
            "lsb"
          ],
          "op": "ret"
        }
      ],
      "name": "lsb",
      "type": "int"
    },
    {
      "args": [
        {
          "name": "a",
          "type": "int"
        },
        {
          "name": "b",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "dest": "zero",
          "op": "const",
          "type": "int",
          "value": 0
        },
        {
          "dest": "one",
          "op": "const",
          "type": "int",
          "value": 1
        },
        {
          "args": [
            "a"
          ],
          "dest": "a_lsb",
          "funcs": [
            "lsb"
          ],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "b"
          ],
          "dest": "b_lsb",
          "funcs": [
            "lsb"
          ],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "a_lsb",
            "b_lsb"
          ],
          "dest": "xor",
          "op": "add",
          "type": "int"
        },
        {
          "args": [
            "xor",
            "one"
          ],
          "dest": "guard",
          "op": "eq",
          "type": "bool"
        },
        {
          "args": [
            "guard"
          ],
          "labels": [
            "if",
            "else"
          ],
          "op": "br"
        },
        {
          "label": "if"
        },
        {
          "args": [
            "one"
          ],
          "op": "ret"
        },
        {
          "label": "else"
        },
        {
          "args": [
            "zero"
          ],
          "op": "ret"
        }
      ],
      "name": "hamm_lsb",
      "type": "int"
    },
    {
      "args": [
        {
          "name": "a",
          "type": "int"
        },
        {
          "name": "b",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "dest": "zero",
          "op": "const",
          "type": "int",
          "value": 0
        },
        {
          "dest": "two",
          "op": "const",
          "type": "int",
          "value": 2
        },
        {
          "dest": "dist",
          "op": "const",
          "type": "int",
          "value": 0
        },
        {
          "label": "while"
        },
        {
          "args": [
            "a",
            "zero"
          ],
          "dest": "a_done",
          "op": "eq",
          "type": "bool"
        },
        {
          "args": [
            "b",
            "zero"
          ],
          "dest": "b_done",
          "op": "eq",
          "type": "bool"
        },
        {
          "args": [
            "a_done",
            "b_done"
          ],
          "dest": "done",
          "op": "and",
          "type": "bool"
        },
        {
          "args": [
            "done"
          ],
          "labels": [
            "end",
            "body"
          ],
          "op": "br"
        },
        {
          "label": "body"
        },
        {
          "args": [
            "a",
            "b"
          ],
          "dest": "hamm",
          "funcs": [
            "hamm_lsb"
          ],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "dist",
            "hamm"
          ],
          "dest": "dist",
          "op": "add",
          "type": "int"
        },
        {
          "args": [
            "a",
            "two"
          ],
          "dest": "a",
          "op": "div",
          "type": "int"
        },
        {
          "args": [
            "b",
            "two"
          ],
          "dest": "b",
          "op": "div",
          "type": "int"
        },
        {
          "labels": [
            "while"
          ],
          "op": "jmp"
        },
        {
          "label": "end"
        },
        {
          "args": [
            "dist"
          ],
          "op": "print"
        },
        {
          "op": "ret"
        }
      ],
      "name": "main"
    }
  ]
}
