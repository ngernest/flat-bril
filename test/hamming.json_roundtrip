{
  "functions": [
    {
      "args": [
        {
          "name": "val",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "dest": "two",
          "op": "const",
          "type": "int",
          "value": 2
        },
        {
          "args": [
            "val",
            "two"
          ],
          "dest": "div",
          "funcs": [],
          "labels": [],
          "op": "div",
          "type": "int"
        },
        {
          "args": [
            "div",
            "two"
          ],
          "dest": "mul",
          "funcs": [],
          "labels": [],
          "op": "mul",
          "type": "int"
        },
        {
          "args": [
            "val",
            "mul"
          ],
          "dest": "lsb",
          "funcs": [],
          "labels": [],
          "op": "sub",
          "type": "int"
        },
        {
          "args": [
            "lsb"
          ],
          "funcs": [],
          "labels": [],
          "op": "ret"
        }
      ],
      "name": "lsb",
      "type": "int"
    },
    {
      "args": [
        {
          "name": "a",
          "type": "int"
        },
        {
          "name": "b",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "dest": "zero",
          "op": "const",
          "type": "int",
          "value": 0
        },
        {
          "dest": "one",
          "op": "const",
          "type": "int",
          "value": 1
        },
        {
          "args": [
            "a"
          ],
          "dest": "a_lsb",
          "funcs": [
            "lsb"
          ],
          "labels": [],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "b"
          ],
          "dest": "b_lsb",
          "funcs": [
            "lsb"
          ],
          "labels": [],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "a_lsb",
            "b_lsb"
          ],
          "dest": "xor",
          "funcs": [],
          "labels": [],
          "op": "add",
          "type": "int"
        },
        {
          "args": [
            "xor",
            "one"
          ],
          "dest": "guard",
          "funcs": [],
          "labels": [],
          "op": "eq",
          "type": "bool"
        },
        {
          "args": [
            "guard"
          ],
          "funcs": [],
          "labels": [
            "if",
            "else"
          ],
          "op": "br"
        },
        {
          "label": "if"
        },
        {
          "args": [
            "one"
          ],
          "funcs": [],
          "labels": [],
          "op": "ret"
        },
        {
          "label": "else"
        },
        {
          "args": [
            "zero"
          ],
          "funcs": [],
          "labels": [],
          "op": "ret"
        }
      ],
      "name": "hamm_lsb",
      "type": "int"
    },
    {
      "args": [
        {
          "name": "a",
          "type": "int"
        },
        {
          "name": "b",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "dest": "zero",
          "op": "const",
          "type": "int",
          "value": 0
        },
        {
          "dest": "two",
          "op": "const",
          "type": "int",
          "value": 2
        },
        {
          "dest": "dist",
          "op": "const",
          "type": "int",
          "value": 0
        },
        {
          "label": "while"
        },
        {
          "args": [
            "a",
            "zero"
          ],
          "dest": "a_done",
          "funcs": [],
          "labels": [],
          "op": "eq",
          "type": "bool"
        },
        {
          "args": [
            "b",
            "zero"
          ],
          "dest": "b_done",
          "funcs": [],
          "labels": [],
          "op": "eq",
          "type": "bool"
        },
        {
          "args": [
            "a_done",
            "b_done"
          ],
          "dest": "done",
          "funcs": [],
          "labels": [],
          "op": "and",
          "type": "bool"
        },
        {
          "args": [
            "done"
          ],
          "funcs": [],
          "labels": [
            "end",
            "body"
          ],
          "op": "br"
        },
        {
          "label": "body"
        },
        {
          "args": [
            "a",
            "b"
          ],
          "dest": "hamm",
          "funcs": [
            "hamm_lsb"
          ],
          "labels": [],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "dist",
            "hamm"
          ],
          "dest": "dist",
          "funcs": [],
          "labels": [],
          "op": "add",
          "type": "int"
        },
        {
          "args": [
            "a",
            "two"
          ],
          "dest": "a",
          "funcs": [],
          "labels": [],
          "op": "div",
          "type": "int"
        },
        {
          "args": [
            "b",
            "two"
          ],
          "dest": "b",
          "funcs": [],
          "labels": [],
          "op": "div",
          "type": "int"
        },
        {
          "args": [],
          "funcs": [],
          "labels": [
            "while"
          ],
          "op": "jmp"
        },
        {
          "label": "end"
        },
        {
          "args": [
            "dist"
          ],
          "funcs": [],
          "labels": [],
          "op": "print"
        },
        {
          "args": [],
          "funcs": [],
          "labels": [],
          "op": "ret"
        }
      ],
      "name": "main"
    }
  ]
}
