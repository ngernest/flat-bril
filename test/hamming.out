@lsb
	op: const		dest: two		type: int		value: 2	
	op: div  		dest: div		type: int		args: ["val", "two"]	
	op: mul  		dest: mul		type: int		args: ["div", "two"]	
	op: sub  		dest: lsb		type: int		args: ["val", "mul"]	
	op: ret  		args: ["lsb"]	
@hamm_lsb
	op: const		dest: zero		type: int		value: 0	
	op: const		dest: one		type: int		value: 1	
	op: call 		dest: a_lsb		type: int		args: ["a"]		funcs: ["lsb"]
	op: call 		dest: b_lsb		type: int		args: ["b"]		funcs: ["lsb"]
	op: add  		dest: xor		type: int		args: ["a_lsb", "b_lsb"]	
	op: eq   		dest: guard		type: bool		args: ["xor", "one"]	
	op: br   		args: ["guard"]		labels: ["if", "else"]
.if
	op: ret  		args: ["one"]	
.else
	op: ret  		args: ["zero"]	
@main
	op: const		dest: zero		type: int		value: 0	
	op: const		dest: two		type: int		value: 2	
	op: const		dest: dist		type: int		value: 0	
.while
	op: eq   		dest: a_done		type: bool		args: ["a", "zero"]	
	op: eq   		dest: b_done		type: bool		args: ["b", "zero"]	
	op: and  		dest: done		type: bool		args: ["a_done", "b_done"]	
	op: br   		args: ["done"]		labels: ["end", "body"]
.body
	op: call 		dest: hamm		type: int		args: ["a", "b"]		funcs: ["hamm_lsb"]
	op: add  		dest: dist		type: int		args: ["dist", "hamm"]	
	op: div  		dest: a 		type: int		args: ["a", "two"]	
	op: div  		dest: b 		type: int		args: ["b", "two"]	
	op: jmp  		labels: ["while"]
.end
	op: print		args: ["dist"]	
	op: ret  	
